transport/src/main/java/io/netty/channel/nio/AbstractNioChannel.java

```diff
@@ -170,16 +170,28 @@ protected final void clearReadPending() {
         if (isRegistered()) {
             EventLoop eventLoop = eventLoop();
             if (eventLoop.inEventLoop()) {
-                readPending = false;
+                clearReadPending0();
             } else {
                 eventLoop.execute(clearReadPendingRunnable);
             }
         } else {
             // Best effort if we are not registered yet clear readPending. This happens during channel initialization.
-            readPending = false;
+            clearReadPending0();
         }
     }
 
+    private void setReadPending0(boolean readPending) {
+        this.readPending = readPending;
+        if (!readPending) {
+            ((AbstractNioUnsafe) unsafe()).removeReadOp();
+        }
+    }
+
+    private void clearReadPending0() {
+        readPending = false;
+        ((AbstractNioUnsafe) unsafe()).removeReadOp();
+    }
```

Comment:
@normanmaurer - Of course this can be done but `clearReadPending` uses a cached `Runnable `clearReadPendingRunnable` and `setReadPending` has been deprecated in favor of `clearReadPending`. If we don't care about `clearReadPending` we can do this.